src/execution_plan/ops/op_var_len_traverse.c

```diff
@@ -0,0 +1,119 @@
+    VarLenTraverse *varLenTraverse = malloc(sizeof(VarLenTraverse));
+    varLenTraverse->g = g;
+    varLenTraverse->A = ae->operands[0].operand;
+    varLenTraverse->srcNode = ae->src_node;
+    varLenTraverse->destNode = ae->dest_node;
+    varLenTraverse->hops = 1;
+    varLenTraverse->minHops = minHops;
+    varLenTraverse->maxHops = maxHops;
+
+    GrB_Matrix_new(&varLenTraverse->frontier, GrB_BOOL, Graph_NodeCount(g), Graph_NodeCount(g));
+
+    // Avoid cycles, add reached nodes to mask.
+    GrB_Matrix_new(&varLenTraverse->mask, GrB_BOOL, Graph_NodeCount(g), Graph_NodeCount(g));
+    GrB_Matrix_dup(&varLenTraverse->mask, varLenTraverse->A);
+
+    /* Setup matrix multiplication descriptor
+     * Mask is complemented, consider [i,j] which were not encountered.
+     * Transpose operand if needed. */
+    GrB_Descriptor_new(&varLenTraverse->desc);
+    GrB_Descriptor_set(varLenTraverse->desc, GrB_MASK, GrB_SCMP);
+    if(ae->operands[0].transpose) GrB_Descriptor_set(varLenTraverse->desc, GrB_INP1, GrB_TRAN);
+
+    /* Perform minimum required hops. */
+    for(int i = 0; i < minHops; i++) _VarLenTraverseOpHop(varLenTraverse);
+    varLenTraverse->it = TuplesIter_new(varLenTraverse->frontier);
+
+    // Set our Op operations
+    OpBase_Init(&varLenTraverse->op);
+    varLenTraverse->op.name = "Variable Length Traverse";
+    varLenTraverse->op.type = OPType_VAR_LEN_TRAVERSE;
+    varLenTraverse->op.consume = VarLenTraverseConsume;
+    varLenTraverse->op.reset = VarLenTraverseReset;
+    varLenTraverse->op.free = VarLenTraverseFree;
+    varLenTraverse->op.modifies = NewVector(char*, 2);
+
+    char *modified = NULL;
+    modified = QueryGraph_GetNodeAlias(qg, *varLenTraverse->srcNode);
+    Vector_Push(varLenTraverse->op.modifies, modified);
+    modified = QueryGraph_GetNodeAlias(qg, *varLenTraverse->destNode);
+    Vector_Push(varLenTraverse->op.modifies, modified);
+
+    return (OpBase*)varLenTraverse;
+}
+
+OpResult VarLenTraverseConsume(OpBase *opBase, QueryGraph* graph) {
+    VarLenTraverse *op = (VarLenTraverse*)opBase;
+    
+    GrB_Index src_id;
+    GrB_Index dest_id;
+
+    while(TuplesIter_next(op->it, &dest_id, &src_id) == TuplesIter_DEPLETED) {
+        // See if we can advance.
+        GrB_Index nvals;
+        GrB_Matrix_nvals(&nvals, op->frontier);
+        if(nvals == 0 || op->hops >= op->maxHops) return OP_DEPLETED;
+        
+        _VarLenTraverseOpHop(op);
+        TuplesIter_reuse(op->it, op->frontier);
+    }
+
+    Node *srcNode = Graph_GetNode(op->g, src_id);
+    Node *destNode = Graph_GetNode(op->g, dest_id);
+    *op->srcNode = srcNode;
+    *op->destNode = destNode;
+    return OP_OK;
+}
+
+OpResult VarLenTraverseReset(OpBase *ctx) {
+    VarLenTraverse *op = (VarLenTraverse*)ctx;
+    op->hops = 1;
+    GrB_free(&op->mask);
+    GrB_Matrix_new(&op->mask, GrB_BOOL, Graph_NodeCount(op->g), Graph_NodeCount(op->g));
+    GrB_Matrix_dup(&op->mask, op->A);
+    TuplesIter_clear(op->it);
```

Comment:
What does this accomplish? The iterator will return DEPLETED on its next invocation, but it does not actually get invoked before the `TuplesIter_reuse` call on 107.